sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/QueryGenerator.kt

```diff
@@ -34,59 +34,65 @@ abstract class QueryGenerator(private val query: BindableQuery) {
   protected fun executeBlock(): CodeBlock {
     val result = CodeBlock.builder()
 
-    val maxIndex = query.arguments.filter { it.type.bindArg?.isArrayParameter() != true }.map { it.index }.max() ?: 0
     val precedingArrays = mutableListOf<String>()
     val bindStatements = CodeBlock.builder()
     val replacements = mutableListOf<Pair<IntRange, String>>()
     val argumentCounts = mutableListOf<String>()
 
     var needsFreshStatement = false
 
-    query.arguments.filterNot { it.type.bindArg?.isArrayParameter() == true }.size.let {
-      if (it != 0) {
-        argumentCounts.add(it.toString())
+    val positionToArgument = mutableListOf<Triple<Int, BindableQuery.Argument, SqlBindExpr?>>()
+    query.arguments.forEach { argument ->
+      if (argument.bindArgs.isNotEmpty()) {
+        argument.bindArgs.forEach { bindArg ->
+          positionToArgument.add(Triple(bindArg.node.textRange.startOffset, argument, bindArg))
+        }
+      } else {
+        positionToArgument.add(Triple(0, argument, null))
```

Comment:
@AlecStrong this part is kinda sketch, but it's only used for inserts in the test and since the sorting is stable, it works. But I'm not thrilled with it. This is only for cases like `foo IN (?, ?)`, which don't have `bindArgs` to find the position with.